### ISSUE TITLE
Some pages require virtual buffer updating in Firefox (AKA <input type="image"> sometimes missing from buffer)

### ISSUE THREAD

**Issue by Cleversn** [nvaccess/nvda#11219]
### Steps to reproduce:
1. Open this page in Firefox:
http://www.mountsinaimonastery.org/support
2. Use the f quick key to see that no form field is present.
3. Press NVDA+f5.
4. Use f or b to see that the Donate Via PayPal button has appeared.
### Actual behavior:
You must guess you need to update the virtual buffer in order for the page to display correctly.
### Expected behavior:
Page should fully load immediately.
#### NVDA installed/portable/running from source:
Installed.
#### NVDA version:
2020.1
#### Windows version:
64-bit Windows 10 pro
#### Name and version of other software in use when reproducing the issue:
Firefox 76.0.1 64-bit.
#### Does the issue still occur after restarting your computer?
Yes.
#### If addons are disabled, is your problem still occuring?
#Yes.
#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
Yes.

**Comment by Adriani90** [comment]
I cannot reproduce this. Is it possible that your focus lands on the button directly when you open the link? Because then NVDA reports not form field. Could you 
1. Open the link
2. Press up arrow
3. Press f?

Does NVDA then report the download button?
In my case I can access the download button everytime when opening the link without having to refresh the virtual buffer.

**Comment by Cleversn** [comment]
Hi, the problem really is true for me. When I open the page, the focus lands at the beginning, unless I had visited the page before, in which case it lands at where I left last time. Then at the place where the button should be, that's just before the line starting with "To cancel your monthly donation", is a blank line. Only when I refresh the buffer, the button appears there.

**Comment by DrSooom** [comment]
I can reproduce it with NVDA 2019.2.1 and Firefox 76.0.1 (64-bit) on Win7x64. This issue could be caused by the external loaded content, because the PayPal image in the button is loaded from https://www.paypalobjects.com/ (see line 567 in the source code).

**Comment by bhavyashah** [comment]
I am unable to replicate this issue with Firefox version 77.0.1 or Brave browser (based on Chrome) version 83.1.10.93 using NVDA version 2020.2 Beta 1. The "Donate with PayPal button" appears straightaway.
@Cleversn  @DrSooom Could you please update to NVDA 2020.2 Beta version and check if the issue persists?

**Comment by Cleversn** [comment]
Hi, in fact, Firefox 77 seems to have corrected the behaviour. Even with NVDA 2020.1, it's all OK now.

**Comment by DrSooom** [comment]
I tested it with Firefox 77.0.1 (64-bit installed) and NVDA 2020.1 Beta 1 (portable) on Win7x64. Here the results:

1. The PayPal button is now accessible loaded in the virtual buffer) after visiting the above mentioned page the very first time or later on again. The last statement means that parts of this website are stored in the cache.
2. And the same result after you reload the page with F5 or CTRL+R.
3. But still not after you reloaded the page including clearing the cache by pressing CTRL+F5.
4. But clearing the whole Firefox cache and visiting the page again the very first time will make the PayPal button accessible, which wasn't the case before.

Thus, just partly fixed.

**Comment by Adriani90** [comment]
Since it is very unlikely that someone will clear the cache with ctrl+f5, and this might cause also other issue as well by the way if you are doing it while you are browsing, I would assess this issue as solved. Note that if you press ctrl+f5 and then restart your browser which is recommended anyway after clearing the cache, then the button is accessible again.
I am closing this issue as works for me.

**Comment by DrSooom** [comment]
> Since it is very unlikely that someone will clear the cache with ctrl+f5, and this might cause also other issue as well […]

@Adriani90: I already though you would say something like this. And no, this isn't imho the point. There is still something working wrong here. And it would be new to me that CTRL+F5 causes problems – as CTRL+F5 should normally fixes problems because it clears obsolete data in the cache and forces to completely load them again from the webserver. Do you have further information about these problems you mentioned? Any sources?

@jcsteh and @MarcoZehe: Your thoughts regarding not rendered external content in the virtual buffer after pressing CTRL+F5 are welcome. Thanks.

**Comment by jcsteh** [comment]
Urgh. This is similar to https://bugzilla.mozilla.org/show_bug.cgi?id=1596007 which I recently fixed, but that one was for the img tag, where this problem is with input type="image". I guess we'll need a similar fix for this case.

For reference, I couldn't reproduce this with the provided URL (probably a faster connection or something), but it's much easier to repro like this:

1. Open this:
    `data:text/html,<input type="image" src="https://via.placeholder.com/100x100.png" alt="blah">`
2. Press control+r to ensure the image isn't cached.
    - Expected: The button should be in the document.
    - Actual: Empty document.

I'll file a Mozilla bug soon.

**Comment by jcsteh** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1647506

**Comment by Cleversn** [comment]
I have just encountered another page with the exact same problem:
1. Open this page:
https://gparted.org/donations.php
2. Press f, no form field;
3. Refresh the virtual buffer with NVDA+f5;
4. Check that now the Paypal button has just appeared;
5. Press F5 to refresh the page;
6. Paypal button has disappeared again. Pressing NVDA+f5 makes it appear again.

NVDA 2020.3, Firefox 81.0.2 64-bit.

**Comment by Adriani90** [comment]
@Cleversn are you still seeing this issue with NVDA 2024.1 Beta 13 and Firefox 123?

**Comment by Cleversn** [comment]
Hi, I cannot reproduce the problem on either page anymore, though the bug that @jcsteh opened at Mozilla appears not fixed yet.

**Comment by Adriani90** [comment]
Given this is not reproducible in NVDA 2024.4.1 with this str and also with the str described in the bug report on Bugzilla, and also not reproducible on earlier versions of NVDA as stated in the last comment above, I am closing as works for me. Tested also in Firefox 132. If this issue still exists for you, please comment and we can reopen.
